File: src/model/models/fun_model_BI1111.jl


## 模块之银行间市场BI1111

##########################################
#状态/开发
##########################################

# module model_BI1111

function model_BI1111!(BB::BankCommercial, BI::BankInterbank, para::Dict, env::Dict)

    env[:indexProcess] = 0 # 初始化过程所在位置
    env[:stateOfSchedule] = :indexing
    @test println("切换调度运作状态为indexing")

    env[:tau] = 0 # 初始化回合

    if (!env[:isModel])
        if (env[:tau] > 0)
            @test println("继续模型model：\n")
        else
            @test println("开始模型model：\n")
        end
    end

    for state in [:indexing, :stepping]
        env[:stateOfSchedule] = state

        ## 过程：银行外部违约损失传染冲击 #BUG测试宏和函数正确性
        env[:processName] = "银行外部违约损失传染冲击过程"
        @scheduler_process BB, BI, env = process_exBank_insolvent!(BB, BI, para, env)

        ## 过程：资不抵债银行间违约损失传染冲击
        env[:processName] = "资不抵债银行间违约损失传染冲击过程"
        @scheduler_process BB, BI, env = process_interBank_insolvent!(BB, BI, para, env)

        ## 过程：外部挤兑流动传染冲击
        env[:processName] = "银行外部挤兑流动传染冲击过程"
        @scheduler_process BB, BI, env = process_exBank_illiquity!(BB, BI, para, env)

        ## 过程：流动性短缺银行间挤兑流动传染冲击
        env[:processName] = "流动性短缺银行间挤兑流动传染冲击过程"
        @scheduler_process BB, BI, env = process_interBank_illiquity!(BB, BI, para, env)

        ## 过程：外生破产银行间挤兑流动传染冲击 #HACK暂时不用
        # env[:processName] = "外生破产银行间挤兑流动传染冲击过程"
        # @run_process BB, BI, env = process_exBank_bankrupt!(BB, BI, para, env)

        ## 过程：破产银行间挤兑流动传染冲击
        env[:processName] = "破产银行间挤兑流动传染冲击过程"
        @scheduler_process BB, BI, env = process_interBank_bankrupt!(BB, BI, para, env)


        # update_B_balanceSheet!(BB, BI,b,ib; byWay = "calc all E_all") # 更新计算各银行之所有者权益
        # BB_tau[env[:tau]] = deepcopy(BB) # 存储该回合传染结果数据
        # BI_tau[env[:tau]] = deepcopy(BI) # 存储该回合传染结果数据
        #TODO 最终破产清算

        ## 判断是否结束
        if !env[:isStep]
            @test println("步进已结束，跳出model_BI1111。")
        end

        if env[:stateOfSchedule] == :standing
            env[:isModel] = false
            env[:isExperiment] = false
        end

        if (!env[:isModel] || !env[:isExperiment])
            @test println("model_BI1111结束。")
        end

    end # for

    return BB, BI, para, env
    # return BB, BI, BB_tau, BI_tau, para, env
end # function

# end # module

File: PySystemicRiskLab/core/template/fun_model_BI1111.py


## 模块之银行间市场IB1111

##########################################
#状态/废弃
##########################################

# module model_IB1111

def model_IB1111(self, BB:BankCommercial, IB:BankInterbank, para:dict, env:dict):

    env['index_process'] = 0 # 初始化过程所在位置
    env['state_of_schedule'] = StateOfScheduleEnum.initializing
    # @testprintln "切换调度状态为$(env['state_of_schedule'])"

    env['tau'] = 0 # 初始化回合

    if (!env['is_model']):
        if (env['tau'] > 0):
            # @testprintln "继续模型model：\n"
        else:
            # @testprintln "开始模型model：\n"
            pass
        pass


    ## 过程：外部资产违约损失传染冲击 #BUG测试宏和函数正确性
    env['process_name'] = "外部资产违约损失传染冲击过程"
    #= @scheduler_process  =#BB, IB, env = processEntity_ExBankInsolvent(BB, IB, paras, env)

    ## 过程：资不抵债银行间违约损失传染冲击
    env['process_name'] = "资不抵债银行间违约损失传染冲击过程"
    #= @scheduler_process  =#BB, IB, env = processEntity_InterBankInsolvent(BB, IB, paras, env)

    ## 过程：外部挤兑流动传染冲击
    env['process_name'] = "银行存款挤兑流动传染冲击过程"
    #= @scheduler_process  =#BB, IB, env = processEntity_ExBankIlliquid(BB, IB, paras, env)

    ## 过程：流动性短缺银行间挤兑流动传染
    env['process_name'] = "流动性短缺银行间挤兑流动传染过程"
    #= @scheduler_process  =#BB, IB, env = processEntity_InterBankIlliquid(BB, IB, paras, env)

    ## 过程：外生破产银行间挤兑流动传染冲击 #HACK暂时不用
    # env['process_name'] = "外生破产银行间挤兑流动传染冲击过程"
    # @execute_branch_node BB, IB, env = process_exBank_bankrupt(BB, IB, paras, env)

    ## 过程：破产银行间挤兑流动传染冲击
    env['process_name'] = "破产银行间挤兑流动传染冲击过程"
    #= @scheduler_process  =#BB, IB, env = processEntity_InterBankBankrupt(BB, IB, paras, env)


    # update_B_balance_sheet(BB, IB,b,ib; by_way = 'calc all E_all') # 更新计算各银行之所有者权益
    # A_data.BB[env['tau']] = deepcopy(BB) # 存储该回合传染结果数据
    # A_data.IB[env['tau']] = deepcopy(IB) # 存储该回合传染结果数据
    #TODO 最终破产清算

    ## 判断是否结束
    if !env['is_step']:
        # @testprintln "步进已结束，跳出model_IB1111。"
        pass

    if env['state_of_schedule'] == StateOfScheduleEnum.idle:
        env['is_model'] = False
        env['is_experiment'] = False
        pass

    if (!env['is_model'] | !env['is_experiment']):
        # @testprintln "model_IB1111结束。"
        pass


    return BB, IB, para, env
    # return BB, IB, A_data.BB, A_data.IB, paras, env
    pass  # method

#     pass # module




